src/node_crypto.cc

```diff
@@ -5159,18 +5156,49 @@ void ECDH::New(const FunctionCallbackInfo<Value>& args) {
   if (!key)
     return env->ThrowError("Failed to create EC_KEY using curve name");
 
-  new ECDH(env, args.This(), std::move(key));
-}
+  ECDH* ecdh = new ECDH(env, args.This(), std::move(key));
 
+  if (args.Length() >= 2) {
+    THROW_AND_RETURN_IF_NOT_BUFFER(env, args[1], "Private key");
 
-void ECDH::GenerateKeys(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
+    BignumPointer priv(BN_bin2bn(
+        reinterpret_cast<unsigned char*>(Buffer::Data(args[1].As<Object>())),
+        Buffer::Length(args[1].As<Object>()),
+        nullptr));
+    if (!priv)
+      return env->ThrowError("Failed to convert Buffer to BN");
 
-  ECDH* ecdh;
-  ASSIGN_OR_RETURN_UNWRAP(&ecdh, args.Holder());
+    if (!ecdh->IsKeyValidForCurve(priv)) {
+      return env->ThrowError("Private key is not valid for specified curve.");
+    }
+
+    int result = EC_KEY_set_private_key(ecdh->key_.get(), priv.get());
+    priv.reset();
```

Comment:
nit: I would put from the declaration of priv to the check on result a couple lines down from here inside a new `{ ... }` block, to force life-time and variable scope to be the same, and prevent accidental use of `priv` later.

Reply:
Sounds like a good idea. This was just moved from `ECDH::SetPrivateKey` so that change would also make sense to apply without the rest of this refactor.